src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -883,333 +986,378 @@ int16_t LoRaWANNode::activateABP(uint8_t initialDr) {
+      out[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] |= RADIOLIB_LORAWAN_FCTRL_ADR_ACK_REQ;
     }
   }
+  
+  // check if we have some MAC commands to append
+  out[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] |= this->fOptsUpLen;
 
   // if the saved confirm-fCnt is set, set the ACK bit
-  bool isConfirmingDown = false;
   if(this->confFCntDown != RADIOLIB_LORAWAN_FCNT_NONE) {
-    isConfirmingDown = true;
-    uplinkMsg[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] |= RADIOLIB_LORAWAN_FCTRL_ACK;
+    out[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] |= RADIOLIB_LORAWAN_FCTRL_ACK;
   }
 
-  LoRaWANNode::hton<uint16_t>(&uplinkMsg[RADIOLIB_LORAWAN_FHDR_FCNT_POS], (uint16_t)this->fCntUp);
+  LoRaWANNode::hton<uint16_t>(&out[RADIOLIB_LORAWAN_FHDR_FCNT_POS], (uint16_t)this->fCntUp);
 
-  // check if we have some MAC commands to append
-  if(fOptsLen > 0) {
-    // assume maximum possible buffer size
-    uint8_t fOptsBuff[RADIOLIB_LORAWAN_FHDR_FOPTS_MAX_LEN];
-
-    this->macQueueToBuff(&(this->commandsUp), fOptsBuff);
-
-    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Uplink MAC payload (%d commands):", this->commandsUp.numCommands);
-    RADIOLIB_DEBUG_PROTOCOL_HEXDUMP(fOptsBuff, fOptsLen);
-
-    uplinkMsgLen = RADIOLIB_LORAWAN_FRAME_LEN(len, fOptsLen);
-    uplinkMsg[RADIOLIB_LORAWAN_FHDR_FCTRL_POS] |= fOptsLen;
+  if(this->fOptsUpLen > 0) {
+    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Uplink MAC payload:");
+    RADIOLIB_DEBUG_PROTOCOL_HEXDUMP(this->fOptsUp, this->fOptsUpLen);
 
     if(this->rev == 1) {
       // in LoRaWAN v1.1, the FOpts are encrypted using the NwkSEncKey
-      processAES(fOptsBuff, fOptsLen, this->nwkSEncKey, &uplinkMsg[RADIOLIB_LORAWAN_FHDR_FOPTS_POS], this->fCntUp, RADIOLIB_LORAWAN_CHANNEL_DIR_UPLINK, 0x01, true);
+      processAES(this->fOptsUp, this->fOptsUpLen, this->nwkSEncKey, &out[RADIOLIB_LORAWAN_FHDR_FOPTS_POS], this->fCntUp, RADIOLIB_LORAWAN_UPLINK, 0x01, true);
     } else {
       // in LoRaWAN v1.0.x, the FOpts are unencrypted
-      memcpy(&uplinkMsg[RADIOLIB_LORAWAN_FHDR_FOPTS_POS], fOptsBuff, fOptsLen);
+      memcpy(&out[RADIOLIB_LORAWAN_FHDR_FOPTS_POS], this->fOptsUp, this->fOptsUpLen);
     }
     
   }
 
   // set the fPort
-  uplinkMsg[RADIOLIB_LORAWAN_FHDR_FPORT_POS(fOptsLen)] = fPort;
+  out[RADIOLIB_LORAWAN_FHDR_FPORT_POS(this->fOptsUpLen)] = fPort;
 
   // select encryption key based on the target fPort
-  uint8_t* encKey = this->appSKey;
-  if(fPort == RADIOLIB_LORAWAN_FPORT_MAC_COMMAND) {
-    encKey = this->nwkSEncKey;
+  uint8_t* encKey;
+  switch(fPort) {
+    case RADIOLIB_LORAWAN_FPORT_MAC_COMMAND:
+      encKey = this->nwkSEncKey;
+      break;
+    case RADIOLIB_LORAWAN_FPORT_PAYLOAD_MIN ... RADIOLIB_LORAWAN_FPORT_PAYLOAD_MAX:
+      encKey = this->appSKey;
+      break;
+    case RADIOLIB_LORAWAN_FPORT_TS009:
+      encKey = this->appSKey;
+      break;
+    case RADIOLIB_LORAWAN_FPORT_TS011:
+      encKey = this->nwkSEncKey;
+      break;
+    default:
+      encKey = this->appSKey;
+      break;
```

Comment:
This can be done with a simpler condition (or a ternary assignment) ;)

```c++
uint8_t* encKey = this->appSKey;
if((fPort == RADIOLIB_LORAWAN_FPORT_MAC_COMMAND) | (fPort == RADIOLIB_LORAWAN_FPORT_TS011)) {
  encKey = this->nwkSEncKey;
}
```